### PR TITLE
feat(stdlib): List.sort via a mergesort (#125)

### PR DESCRIPTION
**Pull request by timgestson**
Implement List.sort with a mergesort algorithm.  Result returns a new list.

### CODE REVIEW

**Comment by timgestson** [comment]
hmm, not quite sure what the failed test here is related to, but can dig deeper

**Comment by spotandjake** [comment]
From what i can tell it looks like this test [here](https://github.com/grain-lang/grain/blob/471df49b2e31a64d08e38bd2ec61744235f14adf/compiler/test/suites/stdlib.re#L12) failed: I think I failed because it looks like they are doing snapshot testing on the generated code. Not 100% sure though hope that helps you, and hopefully a team member can confirm. maybe @ospencer can confirm.

**Comment by phated** [comment]
@spotandjake is correct. Sometimes changes in the stdlib cause snapshots to be updated. You can run the tests with -u to update them

File: stdlib/list.gr [hunk 1]
```diff
@@ -329,3 +329,42 @@ let join = (separator: String, items: List<String>) => {
     Some(s) => s,
   }
 }
+
+/**
+ * Sorts the given list based on a given comparator function via a mergesort algorithm. The resulting list is sorted in increasing order.
+ * 
+ * @param comp: (a, a) -> Number - A function that takes two arguments. It must return 0 if its arguments compare as equal, a positive Number if the first is greater, and a negative Number if the first is smaller
+ * @param list: List<a> - The list to be sorted
+ * @returns List<a>
```
Comment (phated) [inline stdlib/list.gr:338]: @timgestson can you update these the same way that you updated Array.sort? 🙏

File: compiler/test/stdlib/list.test.gr [hunk 2]
```diff
@@ -219,3 +219,15 @@ assert init(-10, idx => fail "Shouldn't be called") == []
...
+// List.sort
+let comp = (left, right) => {
+    match((left, right)) {
+        (left, right) when left > right => 1,
+        (left, right) when left == right => 0,
+        _ => -1
+    }
+}
+assert sort(comp, [3,5,2,4,1]) == [1,2,3,4,5]
+assert sort(comp, []) == []
```
Comment (phated) [inline compiler/test/stdlib/list.test.gr:233]: Do we want any other tests, like the strings test in Array.sort?

**Review by phated — COMMENTED** [review]
A couple of things

File: compiler/test/stdlib/list.test.gr [hunk 3]
```diff
@@ -219,3 +220,24 @@ assert init(-10, idx => fail "Shouldn't be called") == []
...
+let comp = (left, right) => {
+    match((String.length(left), String.length(right))) {
+        (left, right) when left > right => 1,
+        (left, right) when left == right => 0,
+        _ => -1
+    }
+}
+let list = ["a", "abcde", "abc", "ab", "abcd", "a"]
+assert sort(comp, list) == ["a", "a", "ab", "abc", "abcd", "abcde"]
```
Comment (marcusroberts) [inline compiler/test/stdlib/list.test.gr:243]: Looks good!  Just nitpicking, Could we test a couple of edge cases - an empty list, a list with values all the same etc?

Comment (timgestson) [reply compiler/test/stdlib/list.test.gr:243]: What do you have in mind?  There is an empty list test on line 234 and this test has duplicate items in it ("a").  Will be happy to add more if you have a specific case in mind

Comment (timgestson) [reply compiler/test/stdlib/list.test.gr:243]: I'll add 1 more with all identical items

**Review by ospencer — COMMENTED** [review]
This looks great! I have two requests of you—could you add a `List.revAppend` (though check with @phated for naming, he might prefer `reverseAppend`) so we can avoid the `reverse(append(` calls? That'd make this a bit more efficient.

The second one is a bit of a bigger ask—if you take a look at OCaml's implementation, they have a neat trick to avoid the calls to `part`—they pass offsets around. That's a bit more complicated though, so it's fine if we want to hold off on that part and open an issue to implement that 🙂

**Comment by timgestson** [comment]
Yeah let's separate out that into another issue.  I added revAppend for now.  Is this what you are referring to? https://github.com/ocaml/ocaml/blob/trunk/stdlib/list.ml#L340

File: stdlib/list.gr [hunk 4]
```diff
@@ -329,3 +329,54 @@ let join = (separator: String, items: List<String>) => {
     Some(s) => s,
   }
 }
+
+/**
+ * Reverses the list and appends items to the end
```
Comment (ospencer) [inline stdlib/list.gr:334]:
```suggestion
 * Reverses the first list and appends the second list to the end.
```

@phated what do you think about mentioning that it's for efficiency? OCaml describes it as
```
`rev_append l1 l2` reverses `l1` and concatenates it with `l2`. This is equivalent to `(List.rev l1) @ l2`, but `rev_append` is tail-recursive and more efficient.
```

File: stdlib/list.gr [hunk 5]
```diff
@@ -329,3 +329,54 @@ let join = (separator: String, items: List<String>) => {
...
+ * 
+ * @param items: Items to be appended to the reversed list
+ * @param list: The list that is being reversed and appended to
+ * @since v0.4.5
+ */
+let revAppend = (items, list) => { 
+  append(reverse(list), items)
+}
```
Comment (ospencer) [inline stdlib/list.gr:342]: Sorry if I was a bit unclear. I wanted this so we could have an efficient implementation. Usually the main list goes last, but for `append`-like functions it goes in the order they get appended:

```suggestion
 * @param list1: The list to reverse
 * @param list2: The list to append
 * @since v0.4.5
 */
let rec revAppend = (list1, list2) => { 
  match (list1) {
    [hd, ...tl] => revAppend(tl, [hd, ...list2]),
    [] => list2
  }
}
```

We'll have to update the call sites to reverse the arguments as well.

**Comment by ospencer** [comment]
> Yeah let's separate out that into another issue. I added revAppend for now. Is this what you are referring to? https://github.com/ocaml/ocaml/blob/trunk/stdlib/list.ml#L340

Yeah, exactly that! But no worries, we can make it a separate issue 🙂

File: stdlib/list.gr [hunk 6]
```diff
@@ -329,3 +329,57 @@ let join = (separator: String, items: List<String>) => {
...
 }
+
+/**
+ * Reverses the first list and appends the second list to the end.
+ * 
+ * @param list1: The list to reverse
+ * @param list2: The list to append
+ * @since v0.4.5
+ */
+let rec revAppend = (list1, list2) => {
```
Comment (ospencer) [inline stdlib/list.gr:340]: @phated Did you have any thoughts on `revAppend` vs `reverseAppend`? I personally like `revAppend` but `reverseAppend` might be more clear. (But I really like `revAppend`.)

**Review by ospencer — APPROVED** [review]
Thanks @timgestson for making this happen and your patience with us! 🙇

File: stdlib/list.gr (hunk 6 above)
Comment (phated) [reply stdlib/list.gr:340]: No opinion really. You know I like verbosity for clarity, but I don't think `rev` gets confused with anything else (can you think of anything?). If you are **really** like `revAppend` then let's stick with it for now.

Comment (ospencer) [reply stdlib/list.gr:340]: Works for me!

**Review by phated — APPROVED** [review]
Approved! (pending @ospencer doesn't come up with something `rev` can be confused with and we regenerate the docs in this PR).

**Comment by ospencer** [comment]
We realized that we don't want to generate the docs just yet! And I looked up all the words that start with `rev`, and I don't think any of them will get confused with reverse. 🚀
